packages/template-retail-react-app/app/components/product-view/index.jsx

```diff
@@ -151,6 +155,55 @@ const ProductView = forwardRef(
         const isProductASet = product?.type.set
         const isProductABundle = product?.type.bundle
         const errorContainerRef = useRef(null)
+        const [disableButton, setDisableButton] = useState(false)
+        const [customInventoryMessage, setCustomInventoryMessage] = useState('')
+
+        useEffect(() => {
+            let shouldDisableButton = showInventoryMessage
+            if (
+                !shouldDisableButton &&
+                (isProductASet || isProductABundle) &&
+                childProductOrderability
+            ) {
+                // if any of the children are not orderable, it will disable the add to cart button
+                const unavailableChildProductKey = Object.keys(childProductOrderability).find(
+                    (key) => {
+                        return childProductOrderability[key].showInventoryMessage
+                    }
+                )
+                shouldDisableButton = !!unavailableChildProductKey
+                if (unavailableChildProductKey) {
+                    const unavailableChildProduct =
+                        childProductOrderability[unavailableChildProductKey]
+                    let currentInventoryMsg = ''
+                    if (unavailableChildProduct.unfulfillable) {
+                        currentInventoryMsg = intl.formatMessage(
+                            {
+                                defaultMessage: 'Only {stockLevel} left for {productName}!',
+                                id: 'use_product.message.inventory_remaining_for_product'
+                            },
+                            {
+                                stockLevel: unavailableChildProduct.stockLevel,
+                                productName: unavailableChildProduct.productName
+                            }
+                        )
+                    }
+                    if (unavailableChildProduct.isOutOfStock) {
+                        currentInventoryMsg = intl.formatMessage(
+                            {
+                                defaultMessage: 'Out of stock for {productName}',
+                                id: 'use_product.message.out_of_stock_for_product'
+                            },
+                            {productName: unavailableChildProduct.productName}
+                        )
+                    }
+                    setCustomInventoryMessage(currentInventoryMsg)
+                } else {
+                    setCustomInventoryMessage('')
+                }
+            }
+            setDisableButton(shouldDisableButton)
+        }, [showInventoryMessage, childProductOrderability])
```

Comment:
Similarly, we can call `useMemo` here. Something like this:

```suggestion
        const {customInventoryMessage, disableButton} = useMemo(() => {
            let shouldDisableButton = showInventoryMessage
            let currentInventoryMsg = ''

            if (
                !shouldDisableButton &&
                (isProductASet || isProductABundle) &&
                childProductOrderability
            ) {
                // if any of the children are not orderable, it will disable the add to cart button
                const unavailableChildProductKey = Object.keys(childProductOrderability).find(
                    (key) => {
                        return childProductOrderability[key].showInventoryMessage
                    }
                )
                shouldDisableButton = !!unavailableChildProductKey
                if (unavailableChildProductKey) {
                    const unavailableChildProduct =
                        childProductOrderability[unavailableChildProductKey]
                    if (unavailableChildProduct.unfulfillable) {
                        currentInventoryMsg = intl.formatMessage(
                            {
                                defaultMessage: 'Only {stockLevel} left for {productName}!',
                                id: 'use_product.message.inventory_remaining_for_product'
                            },
                            {
                                stockLevel: unavailableChildProduct.stockLevel,
                                productName: unavailableChildProduct.productName
                            }
                        )
                    }
                    if (unavailableChildProduct.isOutOfStock) {
                        currentInventoryMsg = intl.formatMessage(
                            {
                                defaultMessage: 'Out of stock for {productName}',
                                id: 'use_product.message.out_of_stock_for_product'
                            },
                            {productName: unavailableChildProduct.productName}
                        )
                    }
                }
            }

            return {
                customInventoryMessage: currentInventoryMsg,
                disableButton: shouldDisableButton
            }
        }, [showInventoryMessage, childProductOrderability])
```